packages/swapper/src/swappers/thorchain/buildThorTrade/buildThorTrade.ts

```diff
@@ -125,25 +147,40 @@ export const buildTrade = async ({ deps, input }: BuildTradeArgs): Promise<ThorT
         quote: quote as TradeQuote<ThorCosmosSdkSupportedChainId>,
       })
 
-      return {
+      if (maybeTxData.isErr()) return Err(maybeTxData.unwrapErr())
```

Comment:
@0xApotheosis @lmyslinski this is going to be a common pattern, i.e. we're taking a `maybeTxData` and want to return a `Result<ThorTrade<ChainId>, Error>` from the parent function.

is this the standard/ergonomic to do this, i.e. unwrap and rewrap the error?

Reply:
Yeah nah, to maintain a functional flow we want be using the `maybeSomething.match({ ok: ..., err: ... })` pattern here to give us the actual value, rather than early existing with an `Err`. See the example above.

This lets us pattern match on the type then get direct access to the object, rather than unwrapping and re-wrapping every time.

Reply:
Totally agree we could use pattern match instead, though the rationale for manually handling intermediary errors and returning an `Ok()` in most (but not all) places is with the various intermediary `Result`s that can return `Err` types, we may have nested result pattern matches, which may feel like Node.JS callback hell all over again, since we're only using a simple monads lib and nothing fancy like `TaskEither`, `pipe` etc from fp-ts and the like.

Will take a look at all of your comments as well as do a full pass on this diff and use `.match()` wherever it actually makes sense i.e doesn't hinder readability.

Though what @0xdef1cafe was actually referring to is returning a re-wrapped error for error propagation purposes, that's something we've discussed yesterday:

The types are expecting a `Result<T, E>` of a given T type, e.g `ThorTrade<ChainId>` in this case. But when we have an intermediary step failing, we want to return the specific error, so it can be propagated to the parent, which can again return it etc, propagating it all the way up. This means that if we were to return e.g `maybeTxData`, a `Result` monad where `T` is of type `CosmosSignTx | ThorchainSignTx`, we wouldn't conform to the return type of `buildTrade`, hence returning the re-wrapped error instead of the monad.

AFAIK, that's the only way to do error propagation since there's no util doing this. In e.g rust, the `?` operator would take care of propagating the original intermediary error through https://doc.rust-lang.org/book/ch09-02-recoverable-errors-with-result.html#a-shortcut-for-propagating-errors-the--operator but there doesn't seem to be a way to achieve this here

Reply:
As far as this one goes, definitely a good candidate for actually using a match since there are no intermediary values! Note I had type errors when implementing this, which led me to investigate this more and it looks like I may have just found the right way to do error propagation, at least in this case (would work for many intermediate `Result`s as well but potentially more verbose):

https://github.com/sniptt-official/monads/tree/main/lib/result#mapufn-val-t--u-resultu-e

This effectively gives us a callback with the unwrapped result, and allow us to return a new result. Two interesting things here:

1. There is no unwrap/re-wrap needed. The callback arg is already unwrapped, and the result will be auto-rewrapped 🎉 
2. The `Err` is left untouched, meaning we can have our cake (return a different type than the one of the original `Result`) and eat it too (return the original error)

878dcaa3593e8e4bb7b1a040a0c78ddbff7fd0fc

Reply:
Brilliant find with `map<U>(fn: (val: T) => U): Result<U, E>` - that does something close to what I thought `match` was doing - I suspect we'll be using it a lot as we go on this journey.

I'm happy with this pattern, thanks for putting in the extra time to uncover a better practice big @gomesalexandre 🤝